content/2021-05-27-consume-drogue-messages/index.md

```diff
@@ -0,0 +1,251 @@
+along with a few of its pre-requisites. Specifically, we'll be using
+[kubectl] and [func], discussed below, to deploy our event source and
+service. And we'll use [drg] and [HTTPie] to test it.
+
+<!-- more -->
+
+
+# The KafkaSource
+
+We need two things: a knative `KafkaSource` to forward the Drogue
+[CloudEvents], and a knative `Service` to receive them. Because
+drogue-cloud is built atop kafka and knative-eventing, the
+`KafkaSource` custom resource definition (CRD) is already on your
+cluster, along with the topic and bootstrap server listed in the
+following YAML. The only thing you may want to change is the
+`consumerGroup`, to ensure the "sink" receives all messages. If you
+create more than one `KafkaSource` for the same topic, you probably
+want to give each a unique `consumerGroup` name.
+
+~~~yaml
+apiVersion: sources.knative.dev/v1alpha1
+kind: KafkaSource
+metadata:
+  name: drogue-messages
+spec:
+  consumerGroup: some-unique-name
+  bootstrapServers:
+    - kafka-eventing-kafka-bootstrap.knative-eventing.svc:9092
+  topics:
+    - knative-messaging-kafka.drogue-iot.iot-channel
+  sink:
+    ref:
+      apiVersion: serving.knative.dev/v1
+      kind: Service
+      name: cloudevents-rust
+~~~
+
+Assuming you put that YAML in a file named `source.yaml` you would
+apply it like so:
+
+```shell
+kubectl apply -f source.yaml
+```
+
+You can then observe that its status won't become "Ready" until we
+create the `drogue-addict` service, configured as its "sink" above:
+
+```shell
+kubectl get kafkasource
+```
+
+
+# Our Knative Service, built by func
+
+The [func] project aims to simplify the creation of Knative Services
+that respond to [CloudEvents]. It was designed to be run as either a
+self-contained CLI or a [kn plugin]. Unpack the [appropriate
+binary](https://github.com/boson-project/func/releases/latest)
+somewhere on your `$PATH` and run the following to get started:
+
+```shell
+func create drogue-addict && cd drogue-addict
+```
+
+Though it supports an ever-expanding [variety of popular
+languages](https://github.com/boson-project/func/blob/main/docs/guides/developers_guide.md),
+its default is nodejs, which will serve our purposes for now, since
+we're just going to log the event we receive, along with the request
+context, just to show what data we have to work with.  Overwrite the
+entire `index.js` file it generated with the following:
+
+~~~javascript
+/**
+ * Log the drogue event 
+ *
+ * @param {Context} context the invocation context
+ * @param {Object} event the CloudEvent 
+ */
+function handle(context, event) {
+  context.log.info("context");
+  console.log(JSON.stringify(context, null, 2));
+  context.log.info("event");
+  console.log(JSON.stringify(event, null, 2));
+};
+
+module.exports = handle;
+~~~
+
+We need to build our function image. Unfortunately, [func] currently
+requires a docker daemon for this. If you're using [podman], you can
```

Comment:
FWIW, it will always require this unless you want to build your function on the cluster (which is not available yet, but will be someday).